cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -707,18 +688,100 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-// s.mu must be locked when calling handleProcessExit
-func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
-	if ip, ok := p.(*process.Init); ok {
-		// Ensure all children are killed
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
+// handleInitExit processes container init process exits.
+// This is handled separately from non-init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - for a given container, the init process exit MUST be the last exit published
+// This is achieved by:
+// - killing all running container processes
+// - waiting for the container's running exec counter to reach 0, or timeout
+// - removing all remaining processes from s.running so that we s.processExits
+// doesn't publish their exits
+// - publishing the received exec exits, and then publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	// kill all running container processes
+	if err := p.KillAll(s.context); err != nil {
+		log.G(s.context).WithError(err).WithField("id", p.ID()).
+			Error("failed to kill init's children")
+	}
+
+	s.lifecycleMu.Lock()
+	defer s.lifecycleMu.Unlock()
+	numRunningExecs := s.runningExecs[c]
+	if numRunningExecs == 0 {
+		s.handleProcessExit(e, c, p)
+		return
+	}
+
+	events := make(chan int, numRunningExecs)
+	s.execCountSubscribers[c] = events
+
+	go func() {
+		defer func() {
+			s.lifecycleMu.Lock()
+			defer s.lifecycleMu.Unlock()
+			delete(s.execCountSubscribers, c)
+		}()
+
+		// wait for running processes to exit
+		timeout := 20 * time.Millisecond
```

Comment:
I worry this timeout is short enough to plausibly get hit while the system is making forward progress on a heavily thrashing system. Given that the wait unblocks immediately once there are no more execs to wait for, a larger timeout would not delay the delivery of the InitExit event under normal circumstances. As a failsafe for container processes getting wedged in uninterruptible sleep and the like, I can't imagine there would be much issue with having an extra second or two (or thirty) delay.

Reply:
I'm okay with either of those – should we do 1, 2 or 30 seconds? @samuelkarp do you have thoughts on this?